source/common/upstream/outlier_detection_impl.cc

```diff
@@ -282,29 +282,41 @@ DetectorImpl::~DetectorImpl() {
   }
 }
 
-std::shared_ptr<DetectorImpl> DetectorImpl::create(
-    const Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
-    Event::Dispatcher& dispatcher, Runtime::Loader& runtime, TimeSource& time_source,
-    EventLoggerSharedPtr event_logger, Random::RandomGenerator& random) {
+std::shared_ptr<DetectorImpl>
+DetectorImpl::create(Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
+                     Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                     TimeSource& time_source, EventLoggerSharedPtr event_logger,
+                     Random::RandomGenerator& random) {
   std::shared_ptr<DetectorImpl> detector(
       new DetectorImpl(cluster, config, dispatcher, runtime, time_source, event_logger, random));
 
   if (detector->config().maxEjectionTimeMs() < detector->config().baseEjectionTimeMs()) {
     throw EnvoyException(
         "outlier detector's max_ejection_time cannot be smaller than base_ejection_time");
   }
-
   detector->initialize(cluster);
 
   return detector;
 }
 
-void DetectorImpl::initialize(const Cluster& cluster) {
+void DetectorImpl::initialize(Cluster& cluster) {
   for (auto& host_set : cluster.prioritySet().hostSetsPerPriority()) {
     for (const HostSharedPtr& host : host_set->hosts()) {
       addHostMonitor(host);
     }
   }
+
+  if (cluster.healthChecker() != nullptr) {
+    cluster.healthChecker()->addHostCheckCompleteCb(
+        [this](HostSharedPtr host, HealthTransition changed_state) {
+          if (Runtime::runtimeFeatureEnabled(
+                  "envoy.reloadable_features.successful_active_health_check_uneject_host") &&
+              changed_state == HealthTransition::Changed &&
+              host->coarseHealth() == Host::Health::Healthy) {
+            unejectHost(host);
```

Comment:
Will this cause a regular transition from healthy -> unhealthy via active hc cause the outlier detector to count this as an unejection causing the stats to get messed up? Should this only respond to health transitions that clear the outlier check flag?

In general this approach of indirectly responding to a health change via the hc callback seems a bit brittle, though I'm not sure what the better approach would be. Maybe @mattklein123 has any ideas?

Reply:
> Will this cause a regular transition from healthy -> unhealthy via active hc cause the outlier detector to count this as an unejection causing the stats to get messed up? Should this only respond to health transitions that clear the outlier check flag?

I do not think. But we can restrict when only respond to health transitions that clear the outlier check flag to be 100 % safe. I will change it like that.

> In general this approach of indirectly responding to a health change via the hc callback seems a bit brittle, though I'm not sure what the better approach would be. Maybe @mattklein123 has any ideas?

I could not think of any better approach. Let us see if @mattklein123 has any ideas